### ISSUE TITLE
Incorrect type inference after type narrowing

### ISSUE THREAD

**Issue by danpascu** [microsoft/pyright#8575]
**Describe the bug**

I have a variable that is `type[T] | None`. Inside a test like `if var is not None: ...` the variable is now considered to be `type[object]*` instead of `type[T]`.

**Code or Screenshots**

```python
from collections.abc import Mapping
from typing import reveal_type

class Test[T, U]:
    type_map: Mapping[U, type[T]]

    def test(self, value: U) -> None:
        _type = self.type_map.get(value, None)
        reveal_type(_type)      # Type of "_type" is "type[T@Test] | None"
        if _type is not None:
            reveal_type(_type)  # Type of "_type" is "type[object]*"  (I expect this be "type[T@Test]")
```

The comments indicate what is happening. I expect `_type` to be of type `type[T@Test]` inside the if statement.

**VS Code extension or command-line**

Command line pyright, version 1.1.373.

**Comment by danpascu** [comment]
Something to note is that it infers `type[object]*` because `T` is unbound. When `T` is bound, it uses the bounded type instead of object. For example for `class Test[T: Base, U]: ...` the inferred type will be `type[Base]*`.

**Comment by erictraut** [comment]
I think the current behavior is technically correct, since `type[object]*` is effectively an intersection between `type[object]` and `type[T]`. This is equivalent to just `type[T]`, but I can understand why this result would be confusing.

I've updated the heuristic for the `T is None` type narrowing logic to elide the intersection in this case, since it can be simplified to just `type[T]`. This will be included in the next release.

**Comment by danpascu** [comment]
Thank you. This is much appreciated.

While I agree that it's technically correct, it's throwing me a bit off every time I see it as I have to double check I didn't make some mistake. With `type[T]` it just instantly clicks and raises no doubts about what is going on.

Also BTW, this was also inconsistent. This slightly extended example below shows that is some cases it didn't exhibit the behavior mentioned above (as can be seen with the `adapter` variable):

```python

from collections.abc import Mapping
from typing import Protocol, reveal_type


class DataAdapter[T](Protocol):
    @staticmethod
    def validate(value: T) -> T:
        ...


class Test[T, U]:
    type_map: Mapping[U, type[T]]
    adapter_map: Mapping[U, type[DataAdapter[T]]]

    def test(self, value: U) -> None:
        _type = self.type_map.get(value, None)
        reveal_type(_type)      # Type of "_type" is "type[T@Test] | None"
        if _type is not None:
            reveal_type(_type)  # Type of "_type" is "type[object]*"  (shouldn't this be "type[T@Test]")
        adapter = self.adapter_map.get(value, None)
        reveal_type(adapter)      # Type of "adapter" is "type[DataAdapter[T@Test]] | None"
        if adapter is not None:
            reveal_type(adapter)  # Type of "adapter" is "type[DataAdapter[T@Test]]"
```

**Comment by danpascu** [comment]
I'm not sure if your fix covers this, so I'll throw it here just in case you need to make some adjustments before the release. While playing around with some things I ran into a weird case related to this issue. See the last `reveal_type` in the example below where the type is unexpectedly put back together after some conditional type narrowing:

```python


from collections.abc import Mapping
from typing import reveal_type


class Test[T, U]:
    type_map: Mapping[U, type[T]]
    fallback_type: type[T] | None

    def func(self, key: U) -> None:
        reveal_type(self.fallback_type)      # Type of "self.fallback_type" is "type[T@Test] | None"
        if key in self.type_map:
            pass
        else:
            if self.fallback_type is None:
                raise TypeError
            reveal_type(self.fallback_type)  # Type of "self.fallback_type" is "type[object]*"
        reveal_type(self.fallback_type)      # Type of "self.fallback_type" is "type[T@Test] | type[object]* | None"
```

That last `type[T@Test] | type[object]* | None` looks really weird and unexpected, even though it's technically correct, but it's definitely redundant and confusing.

**Comment by erictraut** [comment]
I've confirmed that my fix gives you the results you are hoping to see with the last two code samples.

**Comment by danpascu** [comment]
Thank you

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.374.
